### PR TITLE
Add Metrics UpDownCounter

### LINKED ISSUE #63648: Support Metrics UpDownCounter instrument
# Background and motivation

In .NET 6.0 we have exposed the Metric APIs with exposing different instrument types `Counter`, `Histogram`, `ObservableCounter`, and `ObservableGauge`. We didn't expose the remaining instruments `UpDownCounter` and `ObservableUpDownCounter`. We were collecting feedback to see the demand and the scenarios needed for such instruments before exposing it. Also, we were waiting to have OpenTelemetry specs marked as stable too.

Now we have the feedback, and we are seeing the demand for such instruments and the scenarios need such instruments. The proposal here is to expose these instruments.

[OpenTelemetry specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#updowncounter) 
https://github.com/open-telemetry/opentelemetry-dotnet/issues/2362

# APIs Proposal

UpDownCounter and ObservableUpDownCounter support only the following generic parameter types: `Byte`, `Int16`, `Int32`, `Int64`, `Single`, `Double`, and `Decimal`.

```C#
namespace System.Diagnostics.Metrics
{
    public sealed class UpDownCounter<T> : Instrument<T> where T : struct
    {
        public void Add(T delta) {  throw null; }
        public void Add(T delta, KeyValuePair<string, object?> tag)  {  throw null; }
        public void Add(T delta, KeyValuePair<string, object?> tag1, KeyValuePair<string, object?> tag2)  {  throw null; }
        public void Add(T delta, KeyValuePair<string, object?> tag1, KeyValuePair<string, object?> tag2, KeyValuePair<string, object?> tag3)  {  throw null; }
        public void Add(T delta, ReadOnlySpan<KeyValuePair<string, object?>> tags) {  throw null; }
        public void Add(T delta, params KeyValuePair<string, object?>[] tags) {  throw null; }
        public void Add(T delta, in TagList tagList) { throw null; }
    }

    public sealed class ObservableUpDownCounter<T> : ObservableInstrument<T> where T : struct
    {
        protected override IEnumerable<Measurement<T>> Observe() { throw null;}
    }

    public class Meter : IDisposable
    {
        public UpDownCounter<T> CreateUpDownCounter<T>(string name, string? unit = null, string? description = null) where T : struct  { throw null; }
        
        public ObservableUpDownCounter<T> CreateObservableUpDownCounter<T>(
                            string name,
                            Func<T> observeValue,
                            string? unit = null,
                            string? description = null) where T : struct { throw null; }
        public ObservableUpDownCounter<T> CreateObservableUpDownCounter<T>(
                            string name,
                            Func<Measurement<T>> observeValue,
                            string? unit = null,
                            string? description = null) where T : struct { throw null; }
        public ObservableUpDownCounter<T> CreateObservableUpDownCounter<T>(
                            string name,
                            Func<IEnumerable<Measurement<T>>> observeValues,
                            string? unit = null,
                            string? description = null) where T : struct { throw null; }        
    }
}
```

# APIs Usage

```C#

static Meter s_meter = new Meter("MyLibrary.Queues", "1.0.0");

static UpDownCounter<int> s_queueSize = s_meter.CreateUpDownCounter<int>("Queue-Size");

static ObservableUpDownCounter<int> s_pullQueueSize = s_meter.CreateObservableUpDownCounter<int>("Queue-Size", () => s_pullQueueSize);

...

s_queueSize.Add(10);
s_queueSize.Add(-2);

```

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes https://github.com/dotnet/runtime/issues/63648

### CODE REVIEW

**Comment by tarekgh** [comment]
CC @reyang @cijothomas @noahfalk

File: src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs [hunk 1]
```diff
@@ -317,6 +317,18 @@ public sealed class Counter<T> : Instrument<T> where T : struct
         internal Counter(Meter meter, string name, string? unit, string? description) :
                         base(meter, name, unit, description) {  throw null; }
     }
+    public sealed class UpDownCounter<T> : Instrument<T> where T : struct
+    {
+        public void Add(T delta) {  throw null; }
```
Comment (reyang) [inline src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:322]: curious, is there a reason to have two spaces after `{` while one space before `}`?

**Review by reyang — APPROVED** [review]

File: src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs (hunk 1 above)
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/ref/System.Diagnostics.DiagnosticSourceActivity.cs:322]: No reason. I'll remove the extra space.

**Comment by tarekgh** [comment]
The CI failure is unrelated and tracked with the issue https://github.com/dotnet/runtime/issues/64389

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ObservableUpDownCounter.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// ObservableUpDownCounter is a metrics observable Instrument which reports increasing or decresing value(s) when the instrument is being observed.
+    /// e.g. process heap size and approximate number of items in a lock-free circular buffer.
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ObservableUpDownCounter.cs:10]:
```suggestion
    /// e.g. the process heap size or the approximate number of items in a lock-free circular buffer.
```

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ObservableUpDownCounter.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    /// Use Meter.CreateObservableUpDownCounter methods to create such ObservableUpDownCounter object.
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/ObservableUpDownCounter.cs:11]:
```suggestion
    /// Use Meter.CreateObservableUpDownCounter methods to create an ObservableUpDownCounter object.
```

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/UpDownCounter.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    {
+        internal UpDownCounter(Meter meter, string name, string? unit, string? description) : base(meter, name, unit, description)
+        {
+            Publish();
+        }
+
+        /// <summary>
+        /// Record the delta value of the measurement. The delta can be positive, negative or zero.
+        /// </summary>
+        /// <param name="delta">The amount to be added, can be positive, negative or zero.</param>
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/UpDownCounter.cs:28]:
```suggestion
        /// <param name="delta">The amount to be added which can be positive, negative or zero.</param>
```

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/UpDownCounter.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        public void Add(T delta) => RecordMeasurement(delta);
+
+        /// <summary>
+        /// Record the delta value of the measurement. The delta can be positive, negative or zero.
+        /// </summary>
+        /// <param name="delta">The amount to be added, can be positive, negative or zero.</param>
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/UpDownCounter.cs:34]:
```suggestion
        /// <param name="delta">The amount to be added which can be positive, negative or zero.</param>
```

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/UpDownCounter.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        /// <param name="tag">A key-value pair tag associated with the measurement.</param>
+        public void Add(T delta, KeyValuePair<string, object?> tag) => RecordMeasurement(delta, tag);
+
+        /// <summary>
+        /// Record the delta value of the measurement. The delta can be positive, negative or zero.
+        /// </summary>
+        /// <param name="delta">The amount to be added, can be positive, negative or zero.</param>
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/UpDownCounter.cs:41]:
```suggestion
        /// <param name="delta">The amount to be added which can be positive, negative or zero.</param>
```

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/UpDownCounter.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        /// <param name="tag1">A first key-value pair tag associated with the measurement.</param>
+        /// <param name="tag2">A second key-value pair tag associated with the measurement.</param>
+        public void Add(T delta, KeyValuePair<string, object?> tag1, KeyValuePair<string, object?> tag2) => RecordMeasurement(delta, tag1, tag2);
+
+        /// <summary>
+        /// Record the delta value of the measurement. The delta can be positive, negative or zero.
+        /// </summary>
+        /// <param name="delta">The amount to be added, can be positive, negative or zero.</param>
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/UpDownCounter.cs:49]:
```suggestion
        /// <param name="delta">The amount to be added which can be positive, negative or zero.</param>
```

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/UpDownCounter.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        /// <param name="tag1">A first key-value pair tag associated with the measurement.</param>
+        /// <param name="tag2">A second key-value pair tag associated with the measurement.</param>
+        /// <param name="tag3">A third key-value pair tag associated with the measurement.</param>
+        public void Add(T delta, KeyValuePair<string, object?> tag1, KeyValuePair<string, object?> tag2, KeyValuePair<string, object?> tag3) => RecordMeasurement(delta, tag1, tag2, tag3);
+
+        /// <summary>
+        /// Record the delta value of the measurement. The delta can be positive, negative or zero.
+        /// </summary>
+        /// <param name="delta">The amount to be added, can be positive, negative or zero.</param>
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/UpDownCounter.cs:58]:
```suggestion
        /// <param name="delta">The amount to be added which can be positive, negative or zero.</param>
```

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/UpDownCounter.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        /// <param name="tags">A span of key-value pair tags associated with the measurement.</param>
+        public void Add(T delta, ReadOnlySpan<KeyValuePair<string, object?>> tags) => RecordMeasurement(delta, tags);
+
+        /// <summary>
+        /// Record the delta value of the measurement. The delta can be positive, negative or zero.
+        /// </summary>
+        /// <param name="delta">The amount to be added, can be positive, negative or zero.</param>
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/UpDownCounter.cs:65]:
```suggestion
        /// <param name="delta">The amount to be added which can be positive, negative or zero.</param>
```

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/UpDownCounter.cs [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        /// <param name="tags">A list of key-value pair tags associated with the measurement.</param>
+        public void Add(T delta, params KeyValuePair<string, object?>[] tags) => RecordMeasurement(delta, tags.AsSpan());
+
+        /// <summary>
+        /// Record the delta value of the measurement. The delta can be positive, negative or zero.
+        /// </summary>
+        /// <param name="delta">The amount to be added, can be positive, negative or zero.</param>
```
Comment (noahfalk) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/UpDownCounter.cs:72]:
```suggestion
        /// <param name="delta">The amount to be added which can be positive, negative or zero.</param>
```

**Review by noahfalk — APPROVED** [review]
LGTM

**Pull request merged**